private/buf/bufgen/bufgen.go

```diff
@@ -44,8 +30,16 @@ const (
 	V1Version = "v1"
 	// V1Beta1Version is the string used to identify the v1beta1 version of the generate template.
 	V1Beta1Version = "v1beta1"
+	// V2Version is the string used to identify the v2 version of the generate template.
+	V2Version = "v2"
 )
 
+// ExternalConfigVersion defines the subset of all config
+// file versions that is used to determine the configuration version.
+type ExternalConfigVersion struct {
```

Comment:
Helpers that should only be used by `bufgenv1` and `bufgenv2`, but not outside of these packages should be in `bufgen/internal` - is this used outside of `bufgen/*` or is this just an internal implementation detail?

Reply:
`ExternalConfigVersion` is also used in [`bufmigrate`](https://github.com/bufbuild/buf/blob/main/private/buf/bufmigrate/v1beta1_migrator.go#L121). With some dependency injection (`Logger`, `ReadWriteBucket`, `ConfigDataProvider`) into `v1beta1Migrator`, we might be able to hide `ExternalConfigVersion` in `internal` and let the migrator call `ReadConfigVersion`. I left it as a TODO so that I can visit this when writing migrator to v2.

After inspecting `bufgen.go`, only `Strategy` is movable and it's moved to `internal`.
[`ReadFromConfig`](https://github.com/bufbuild/buf/pull/2140/files#diff-855d43fb4faf389939e0de63c9b9eb2a5a7ae640984bf23ded98ff5a2441026aR90) is not moved to `internal`, to avoid circular import, because `ReadFromConfig` has a parameter of type `ConfigDataProvider`, which is defined in `bufgen`.